### ISSUE TITLE
Docker exec hangs on exit when stdout is open

### ISSUE THREAD

**Issue by cfstras** [moby/moby#33039]
**Description**

When `docker exec -it` is called on a running container, and stdout is kept open from a command invoked inside that shell, the exec shell hangs until the command exits. 

**Steps to reproduce the issue:**
```
user@carlyle-linux ~ % docker run -d --name test alpine sh -c 'while true; do sleep 10;done'
60b69398294b2a2490c07cc1e87fa63755e623d56f041817835b646544593e17
user@carlyle-linux ~ % docker exec -it test sh
/ # touch /var/log/test
/ # tail -f /var/log/test &
/ # exit
<hangs>
^Z
^C
<still hangs>
# in another terminal: docker exec test killall tail
# now it exist
user@carlyle-linux ~ %
```
**Describe the results you received:**
`docker exec` closes stdout and `tail` exits when shell is exited

**Describe the results you expected:**
`docker exec` waits for tail to close stdout, which never happens

**Additional information you deem important (e.g. issue happens only occasionally):**
Reproduced so far on docker for-win, multiple versions, and docker Arch Linux. 

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 10
 Running: 3
 Paused: 0
 Stopped: 7
Images: 542
Server Version: 17.04.0-ce
Storage Driver: btrfs
 Build Version: Btrfs v4.10.2
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.10.13-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.64GiB
Name: carlyle-linux
ID: NVQB:<snip>:O2PA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
I think this is a duplicate of https://github.com/moby/moby/issues/26403, and the expected behaviour; the `docker exec` process in interactive mode waits until processes started in the interactive session terminate. In this case, the `tail -f` is started in the background, so won't terminate, thus exec does not detach.

You can probably detach the session with `CTRL-p,q`, but the process will remain running

**Comment by javabrett** [comment]
> the docker exec process in interactive mode waits until processes started in the interactive session terminate.

**TL;DR**: if you background a process in a `docker exec -i` session, make sure you close/redirect its _stdout_ and _stderr_.  If you do that, you should be able to background the process and return from `docker exec -i`.

I have found that it is possible to have the session started by `docker exec -i ...` start a long-running process, and leave that process running, whilst `docker exec` itself terminates.  Even did it with a _heredoc_.

The key I found is that in-addition to backgrounding the job (# runlong &), you need to ensure that both _stdout_ and _stderr_ are closed/redirected, that is, one of e.g.:

```
# runlong >runlong.out 2>runlong.err &
# runlong >/dev/null 2>&1 &
```

In that case, when the `docker exec -i` session reaches _EOF_, it will exit immediately and return to the calling-shell.  If you omit the redirects, it will block and only return if `runlong` eventually exits.

Be especially wary of relying on `nohup` to redirect _stdout_ and _stderr_.  It has a feature I had forgotten about that caught me out - if `nohup` thinks it is not attached to a TTY (i.e. you didn't run `docker exec -t`), it silently refuses to redirect to `nohup.out`, which will result in the hanging-behaviour when `docker exec -i` won't return at _EOF_.

**Comment by rickmanelius** [comment]
We seemed to have hit upon this issue as well (https://github.com/drud/ddev/issues/732#issuecomment-373197070). Posting not to pile on, but merely to indicate that #26403 was not a fix and it's still present in the latest 17.X branch.

**Comment by xyoun** [comment]
that's really a problem. the health check script also hangs. and output from docker inspect:

```
"Health": {
    "Status": "unhealthy",
    "FailingStreak": 7,
    "Log": [
        {
        "Start": "...",
        "End": "...",
        "ExitCode": -1,
        "Output": "Health check exceeded timeout (30s)"
    }
    ]
}
```

**Comment by dlipovetsky** [comment]
Saw this remains open and want to add some information: I am able to reproduce the busybox example with Docker CE 17.09.1, but not with with Docker CE 18.09.0. With 18.09.0, the exit does not hang.
```
$ docker info
Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 17
Server Version: 18.09.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: c4446665cb9c30056f4998ed953e6d4ff22c7c39
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.18.18-200.fc28.x86_64
Operating System: Fedora 28 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.43GiB
Name: <redacted>
ID: <redacted>
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Comment by codingforfun** [comment]
> **Describe the results you received:**
> `docker exec` closes stdout and `tail` exits when shell is exited
> 
> **Describe the results you expected:**
> `docker exec` waits for tail to close stdout, which never happens

@cfstras Don’t you mean the opposite thing?

**Comment by cfstras** [comment]
yes

**Comment by skeetor** [comment]
I had the same problem when I run multiple build jobs on bamboo, I'm executing several debian jessie containers and some debian stretch containers. The stretch always hangs, when I run this in parallel. There is a script running in the container and it hangs on the last line when it already should have exited. I waited for half an hour and then killed it.

We are running Docker version 18.09.3, build 774a1f4
I looked at the release notes of newer version, but didn't find a hint that this is fixed.
The server is running Ubuntu 16.04.6 LTS

**Comment by pratikbalar** [comment]
having same problem with GitHub actions self-hosted runner https://github.com/actions/runner/issues/1214#issuecomment-917349198
